### PR TITLE
daemon: refactor config loading

### PR DESCRIPTION
**Pull request by thaJeztah**
extracting first couple of commits from https://github.com/moby/moby/pull/43703

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
@tianon @corhere ptal

**Review by corhere — APPROVED** [review]

**Comment by thaJeztah** [comment]
Ah, dang; I screwed up the last commit; flipped the check (I wanted to add a check for the correct error). Let me fix that

```
=== FAIL: github.com/docker/docker/daemon/config TestDaemonConfigurationMergeConcurrentError (0.00s)
    config_test.go:100: assertion failed: error is not nil: merged configuration validation from file and command line flags failed: invalid max concurrent downloads: -1
```

File: cmd/dockerd/daemon.go [hunk 1]
```diff
@@ -92,7 +90,7 @@ func (cli *DaemonCli) start(opts *daemonOptions) (err error) {
 
 	if opts.Validate {
 		// If config wasn't OK we wouldn't have made it this far.
-		fmt.Fprintln(os.Stderr, "configuration OK")
+		_, _ = fmt.Fprintln(os.Stderr, "configuration OK")
```
Comment (tianon) [inline cmd/dockerd/daemon.go:93]: :raised_eyebrow:

File: daemon/config/config_test.go [hunk 2]
```diff
@@ -546,7 +554,9 @@ func TestReloadWithDuplicateLabels(t *testing.T) {
 	flags := pflag.NewFlagSet("test", pflag.ContinueOnError)
 	flags.String("config-file", configFile, "")
 	flags.StringSlice("labels", lbls, "")
-	err := Reload(configFile, flags, func(c *Config) {})
+	err := Reload(configFile, flags, func(c *Config) {
+		assert.Check(t, is.DeepEqual(c.Labels, []string{"foo=the-same"}))
```
Comment (tianon) [inline daemon/config/config_test.go:580]: Do we need to validate somehow that this assertion actually fired?  (I guess all the "reloaded when we expected it to" tests above do that?)

File: daemon/daemon.go [hunk 3]
```diff
@@ -1347,14 +1345,6 @@ func (daemon *Daemon) setGenericResources(conf *config.Config) error {
 	return nil
 }
 
-func setDefaultMtu(conf *config.Config) {
-	// do nothing if the config does not have the default 0 value.
-	if conf.Mtu != 0 {
-		return
-	}
-	conf.Mtu = config.DefaultNetworkMtu
```
Comment (tianon) [inline daemon/daemon.go:1355]: Does this technically mean we can no longer set `--mtu=0` or `"mtu":0` in our configuration to receive the default value? :innocent:

(Not sure whether anyone's actually doing that, but the current code would support that, right?  The "workaround" for anyone doing that would just be to remove the flag/value, right?  Unless the flag can override the configuration or something but I think that's an error so this is probably fine.)

**Review by tianon — APPROVED** [review]
Nice, looks good to me!  I had a couple comments, but I don't think any of them are particularly important. :+1:

File: cmd/dockerd/daemon.go (hunk 1 above)
Comment (thaJeztah) [reply cmd/dockerd/daemon.go:93]: Got fed up with my IDE complaining about the unhandled error 😂

File: daemon/config/config_test.go (hunk 2 above)
Comment (thaJeztah) [reply daemon/config/config_test.go:580]: Good point; let me add the `reloaded` bool for this one as well

**Review by tianon — APPROVED** [review]

**Pull request merged**